### PR TITLE
NIP-11 extension for Relay Limitations

### PR DESCRIPTION
**Pull request by doc-hex**
This PR is an extension to NIP-11 and allows the relay to specify its technical, and other, limitations so that clients can adjust their behaviours to better suit the relay.

This includes such things as:
- maximum length of incoming JSON (which has many impacts)
- retention times for specific event types
- political considerations
- content policies

To be very clear: this document DOES NOT propose any values for these limitations. It is merely a way for cooperating clients to know what they are dealing with. All are optional fields.

This PR could be merged into NIP-11 as it is a pure extension to that JSON response, or be its own thing. I'm proposing 111 as its number in that case.

### CODE REVIEW

**Comment by doc-hex** [comment]
Here is the NIP-11 reply for <https://brb.io>

```shell
% curl -sH "Accept: application/nostr+json" https://brb.io | json
```
```json
{
    "name": "BRB.io",
    "description": "Nostr relay server by Coinkite. A work in progress.",
    "pubkey": "3c1ba7d42c873c2f89caf1ca79b4ead6513385de53743fa6eb98c3705655695c",
    "contact": "mailto:brb@coinkite.com",
    "software": "BRB-noserver",
    "version": "0.6",
    "supported_nips": [
        1,
        2,
        4,
        9,
        11,
        12,
        15,
        16,
        20,
        22,
        40
    ],
    "limitation": {
        "max_raw_json": 16384,
        "max_subscriptions": 100,
        "max_filters": 200,
        "max_limit": 5000,
        "subid_length": 100,
        "min_prefix": 4
    },
    "relay_countries": [
        "CA",
        "DE"
    ],
    "retention": {
        "default": null
    }
}
```

**Comment by fiatjaf** [comment]
Looks good at first glance, but the NIP number is overflowing.

**Review by 1440000bytes — COMMENTED** [review]
Everything looks okay except 'Content Limitations'. Not sure if this has to be mentioned in a NIP.

**Comment by doc-hex** [comment]
If you assign a new NIP, I will update and add references needed. If you want it merged into NIP-11, that's also an option. Personally, I think a new NIP number is order.

File: 111.md [hunk 1]
```diff
@@ -0,0 +1,183 @@
+NIP-111
+=======
+
+Addition Relay Information for NIP-11
```
Comment (Cameri) [inline 111.md:4]:
```suggestion
Additional Relay Information for NIP-11
```

File: 111.md [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+-------------------------------------
+
+`draft` `optional` `author:doc-hex`
+
+Relays may provide addition metadata to clients to inform them of
```
Comment (Cameri) [inline 111.md:9]:
```suggestion
Relays may provide additional metadata to clients to inform them of
```

File: 111.md [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+more detailed limitations and policy choices. This is made available
+as additional fields in the a JSON document already defined by
+[NIP-11](11.md). All are optional and may be omitted.
+
+All numbers and values show in theses examples, are merely _examples_
+and do not constitute a default value or expectation. The purpose
```
Comment (Cameri) [inline 111.md:15]:
```suggestion
and do not constitute default values or expectations. The purpose
```

File: 111.md [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+
+These are limitations imposed by the relay on clients. Your client
+should expect that requests which exceed these *practical* limitations
+are rejected or fail immediately.
+
+```json
+{
+...
+  limitation: {
+        max_raw_json: 16384,
```
Comment (Cameri) [inline 111.md:44]:
```suggestion
        max_message_length: 16384,
```

File: 111.md [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+        max_filters: 100,
+        max_limit: 5000,
+        subid_length: 100,
+        min_prefix: 4,
+  }
+...
+}
+```
+
+- `max_raw_json`: this is the maximum number of bytes for incoming JSON that the relay
```
Comment (Cameri) [inline 111.md:55]:
```suggestion
- `max_message_length`: this is the maximum number of bytes for incoming JSON that the relay
```

File: 111.md [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+will attempt to decode and act upon. When you send large subscriptions, you will be
+limited by this value. It also effectively limits the maximum size of any event. Value is
```
Comment (Cameri) [inline 111.md:57]:
```suggestion
will attempt to decode and act upon. When you send large subscriptions or messages,
you will be limited by this value. It also effectively limits the maximum size of any event. It is
```

File: 111.md [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+
+- `max_subscriptions`: total number of subscriptions that may be
+active on a single websocket connection to this relay. It's possible
+that authenticated clients with a (paid) relationship to the relay
+may have higher limits.
+
+- `max_filters`: maximum number of filter values in each subscription.
+Must be one or higher.
+
+- `subid_length`: maximum length of subscription id as a string.
```
Comment (Cameri) [inline 111.md:69]:
```suggestion
- `max_subid_length`: maximum length of subscription id as a string.
```

File: 111.md [hunk 8]
```diff
@@ -0,0 +1,183 @@
...
+
+```json
+{
+...
+  limitation: {
+        max_raw_json: 16384,
+        max_subscriptions: 20,
+        max_filters: 100,
+        max_limit: 5000,
+        subid_length: 100,
```
Comment (Cameri) [inline 111.md:48]:
```suggestion
        max_subid_length: 100,
```

File: 111.md [hunk 9]
```diff
@@ -0,0 +1,183 @@
...
+```json
+{
+...
+  retention: {
+    default: 3600,
+    by_kind: { 0: null, 2: null, 3: null, 4: 172800 },
+  }
+...
+}
+```
```
Comment (Cameri) [inline 111.md:102]:
```suggestion
```json
{
...
  retention: [
    { kinds: [0, 1, [5,7], [40,49]], time: 1000 },
    { kinds: [[40000, 49999], time: 100 },
    { kinds: [[30000, 39999], count: 1000 },
    { time: 3600, count: 10000 }
  ]
...
}
```
```

I think this way it is more flexible and one can specify kind ranges, time and count.

File: 111.md [hunk 10]
```diff
@@ -0,0 +1,183 @@
...
+For public text notes at least, a relay may try to foster a
+local community. This would encourage users to follow the global
+feed on that relay, in addition to their usual individual follows.
+To support this goal, relays MAY specify some of the following values.
+
+```json
+{
+...
+  language_tags: [ 'en', 'en-419' ],
+  only_sfw: True,
```
Comment (Cameri) [inline 111.md:164]:
```suggestion
  tags: ["sfw-only", "anime", "bitcoin-only"],
```
We can't anticipate all possible community preferences. Let's turn this into tags.

**Comment by Cameri** [comment]
@doc-hex should we add a setting for specifying whether the relay is `private` or whether it requires `pow` or `payment`?

**Comment by nvk** [comment]
> @doc-hex should we add a setting for specifying whether the relay is `private` or whether it requires `pow` or `payment`?

+1

**Comment by fiatjaf** [comment]
I think this should just be shoved into NIP-11. Why not?

**Comment by Cameri** [comment]
> I think this should just be shoved into NIP-11. Why not?

yeah why not? it won't break any client and 1 less NIP

File: 111.md (hunk 2 above)
Comment (doc-hex) [reply 111.md:9]: yes

File: 111.md (hunk 4 above)
Comment (doc-hex) [reply 111.md:44]: sure

File: 111.md (hunk 6 above)
Comment (doc-hex) [reply 111.md:57]: kept orig because that's where I see the issue as a relay operator

File: 111.md (hunk 7 above)
Comment (doc-hex) [reply 111.md:69]: yes

File: 111.md (hunk 9 above)
Comment (doc-hex) [reply 111.md:102]: yes. this is superior. will apply

**Comment by doc-hex** [comment]
> @doc-hex should we add a setting for specifying whether the relay is `private` or whether it requires `pow` or `payment`?

adding these limitations:

> - `min_pow_difficulty`: new events will require at least this difficulty of PoW,
> based on [NIP-13](13.md), or they will be rejected by this server.
> 
> - `auth_required`: this relay requires [NIP-42](42.md) authentication 
> to happen before a new connection may perform any other action.
> Even if set to False, authentication may be required for specific actions.

File: 111.md (hunk 6 above)
Comment (Cameri) [reply 111.md:57]: As a relay operator there's also issues with big events with lots of content or tags. You should reconsider. Check out https://nostr.io/stats for average event sizes.

File: 111.md (hunk 9 above)
Comment (Cameri) [reply 111.md:102]: I missed a few ] closing brackets there oops.

File: 111.md (hunk 6 above)
Comment (doc-hex) [reply 111.md:59]: yeah, I'm adding those 2

Comment (doc-hex) [reply 111.md:61]: see 6b3038f..27ba498

**Comment by doc-hex** [comment]
Updated server response example:

```
% curl -sH "Accept: application/nostr+json" https://brb.io | json
```

```
{
    "name": "BRB.io",
    "description": "Nostr relay server by Coinkite. A work in progress.",
    "pubkey": "3c1ba7d42c873c2f89caf1ca79b4ead6513385de53743fa6eb98c3705655695c",
    "contact": "mailto:brb@coinkite.com",
    "software": "BRB-noserver",
    "version": "0.6",
    "supported_nips": [
        1,
        2,
        4,
        9,
        11,
        12,
        15,
        16,
        20,
        22,
        40
    ],
    "limitation": {
        "max_message_length": 16384,
        "max_content_length": 12000,
        "max_subscriptions": 100,
        "max_filters": 200,
        "max_limit": 5000,
        "max_subid_length": 100,
        "min_prefix": 4,
        "max_event_tags": 100,
        "auth_required": false
    },
    "relay_countries": [
        "CA",
        "DE"
    ],
    "tags": [
        "bitcoin-only"
    ],
    "retention": []
}
```

File: 111.md [hunk 11]
```diff
@@ -0,0 +1,212 @@
...
+        max_subscriptions: 20,
+        max_filters: 100,
+        max_limit: 5000,
+        max_subid_length: 100,
+        min_prefix: 4,
+        max_event_tags: 100,
+        max_content_length: 8196,
+        min_pow_difficulty: 30,
+        auth_required: True,
+  }
```
Comment (Cameri) [inline 111.md:54]:
```suggestion
        auth_required: true,
        payment_required: true,
  }
```

File: 111.md [hunk 12]
```diff
@@ -0,0 +1,212 @@
...
+subject to the `max_message_length`, if defined.
+
+- `min_pow_difficulty`: new events will require at least this difficulty of PoW, 
+based on [NIP-13](13.md), or they will be rejected by this server.
+
+- `auth_required`: this relay requires [NIP-42](42.md) authentication
+to happen before a new connection may perform any other action.
+Even if set to False, authentication may be required for specific actions.
+
+
```
Comment (Cameri) [inline 111.md:98]:
```suggestion
- `payment_required`: this relay requires payment before a new connection may perform any action.

```

File: 111.md [hunk 13]
```diff
@@ -0,0 +1,212 @@
...
+- `posting_policy` is a link to a human-readable page which specifies the
+  community policies for the relay. In cases where `only_sfw` is True, it's
+  important to link to a page which gets into the specifics of your posting policy.
+
+The `description` field should be used to describe your community
+goals and values, in brief. The `posting_policy` is for additional
+detail and legal terms. Use the `tags` field to signify limitations
+on content, or topics to be discussed, which could be machine
+processed by appropriate client software.
+
```
Comment (Cameri) [inline 111.md:212]:
```suggestion
Pay-To-Relay
==========

Relays that require payments may want to expose their fee schedules.

```json
{
...
  payments_url: "https://my-relay/payments",
  fees: {
    "admission": [{ amount: 1000000, unit: 'msats' }],
    "subscription": [{ amount: 5000000, unit: 'msats', period: 2592000 }],
    "publication": [{ kinds: [4], amount: 100, unit: 'msats' }],
  },
...
}
```
- `payments_url` is an URL that points to a Web page where the use can go to pay for different fees. Used by clients that do not support seamless pay-to-relay integration.

- `fees` is an object containing various types of fees charged by the relay. The value of each type of fee is an array containing one or more objects with the following suggested fields:
  - amount: amount charged as a number
  - unit: unit of account (e.g. msats, sats, BTC)
  - period: how often this fee is charged
  - kinds: limits which event kinds this fee is applicable to.
```

File: 111.md [hunk 14]
```diff
@@ -0,0 +1,212 @@
...
+  For example `sfw-only` indicates hat only "Safe For Work" content
+  is encouraged on this relay. This relies on assumptions of what the
+  "work" "community" feels "safe" talking about. In time, a common
+  set of tags may emerge that allow users to find relays that suit
+  their needs, and client software will be able to parse these tags easily.
+  The `bitcoin-only` tag indicates that any *altcoin*, *"crypto"* or *blockchain*
+  comments will be ridiculed without mercy.
+
+- `posting_policy` is a link to a human-readable page which specifies the
+  community policies for the relay. In cases where `only_sfw` is True, it's
```
Comment (Cameri) [inline 111.md:204]:
```suggestion
  community policies for the relay. In cases where `sfw-only` is True, it's
```

**Comment by fiatjaf** [comment]
I think it will be less confusing for everybody if we just add a section "extra fields" to NIP-11 and keep adding things there as required.

**Comment by mikedilger** [comment]
# Some of the text at the end is really big. I'm thinking it was not all meant to be a section title

**Comment by KiPSOFT** [comment]
> 

In my opinion, as new additions are made to the extra_fields field, this field will become increasingly complex. If we are to continue with the same NIP (NIP-11), then I think @doc-hex proposal is a cleaner solution. There are no breaking changes here.

File: 111.md [hunk 15]
```diff
@@ -0,0 +1,229 @@
...
+- `posting_policy` is a link to a human-readable page which specifies the
+  community policies for the relay. In cases where `sfw-only` is True, it's
+  important to link to a page which gets into the specifics of your posting policy.
+
+The `description` field should be used to describe your community
+goals and values, in brief. The `posting_policy` is for additional
+detail and legal terms. Use the `tags` field to signify limitations
+on content, or topics to be discussed, which could be machine
+processed by appropriate client software.
+Pay-To-Relay
```
Comment (Cameri) [inline 111.md:214]:
```suggestion
processed by appropriate client software.

Pay-To-Relay
```

File: 111.md [hunk 16]
```diff
@@ -0,0 +1,229 @@
...
+{
+...
+  payments_url: "https://my-relay/payments",
+  fees: {
+    "admission": [{ amount: 1000000, unit: 'msats' }],
+    "subscription": [{ amount: 5000000, unit: 'msats', period: 2592000 }],
+    "publication": [{ kinds: [4], amount: 100, unit: 'msats' }],
+  },
+...
+}
```
Comment (Cameri) [inline 111.md:229]:
```suggestion
}

Alternate Relay URLs
================

Some relays expose alternate URLs such as onions or plan IP addresses. Relays can publicly expose additional URLs that can be used to reach the same relay instance.

```json
{
  "alternate_relay_urls": ["ws://some-alternate-url.onion", "ws://12.34.56.78:7000"]
}
```

```

File: 111.md [hunk 17]
```diff
@@ -0,0 +1,212 @@
...
+Support for this NIP should be documented in the `supported_nips` list
+of the same response.
+
+```json
+{
+...
+  supported_nips: <a list of NIP numbers supported by the relay, including this one>,
+...
+}
+```
```
[the remark this answers is not in the document]
Comment (kappaseijin) [reply 111.md:31]: I have three questions. 
New NIP's name is "NIP-11a". But md file name is yet ''111.md".
By the way, A part of "Relay Information Document" of NIP-11:
`"supported_nips": <a list of NIP numbers supported by the relay>,`
https://github.com/nostr-protocol/nips/blob/master/11.md#relay-information-document
"supported_nips" is a list of numbers. not a list of strings.

Question 1.
Will the file name "111.md" be changed to "11a.md"?

Question 2.
Is "NIP-11a" just a specification name, or does "supported_nips" contain any of the values 11 , 111, or "11a"?

Question 3.
Will both "supported_nips" and "supported_nip_extensions" be included in the JSON?

**Comment by kappaseijin** [comment]
@Semisol  @fiatjaf Please give me a reaction my questions.
https://github.com/nostr-protocol/nips/pull/163#discussion_r1106934867

**Comment by kappaseijin** [comment]
I got the answers to [my questions](https://github.com/nostr-protocol/nips/pull/163#discussion_r1106934867). 
Nostream showed me a prior implementation. I see that relays around the world have already adopted it.
I have no more question on this PR. Thank you very much.
Hope to have it merged in the near future.
https://github.com/Cameri/nostream/commit/b9a13c886811f8ab1c8a479783905dc16cab6959
https://snowcait.github.io/nostr-relays/

**Comment by fiatjaf** [comment]
Who volunteers to take this stuff here and shove it into NIP-11 directly?

**Comment by staab** [comment]
Looks like this was merged indirectly? Closing.

**Pull request closed without merging**